### PR TITLE
WithOptions<T> loses CommandTimeout property

### LINKED ISSUE #4476: WithOptions<T> looses CommandTimeout property
### Describe your issue

Seems that configuring context with `AddLinqToDBContext` has two ways to set `DataContextOptions`. One of which looses parameters.

```c#
using System.Diagnostics;
using Microsoft.Extensions.DependencyInjection;
using LinqToDB;
using LinqToDB.Data;
using LinqToDB.AspNet;
using LinqToDB.DataProvider.SqlServer;

namespace ConsoleApp1;

internal sealed class DbContext : DataConnection
{
    public DbContext(DataOptions options) : base(options)
    {
        Debug.Assert(options.DataContextOptions.CommandTimeout == 91); // always ok
        Debug.Assert(CommandTimeout == 91); // failing
    }
}

static class Program
{
    static void Main()
    {
        var services = new ServiceCollection();

        services.AddLinqToDBContext<DbContext>((provider, options) =>
        {
            return options
                .UseSqlServer("some conn string", SqlServerVersion.v2019, SqlServerProvider.MicrosoftDataSqlClient)
                .WithOptions<DataContextOptions>(o => o with { CommandTimeout = 91 }) // this failing
                // .WithOptions(options.DataContextOptions with { CommandTimeout = 91 }) // this works
                ;
        });

        var container = services.BuildServiceProvider();

        container.GetService<DbContext>();

        System.Console.WriteLine("Ok");
    }
}
```

### Environment details

`Linq To DB` version: 5.4.1

Database (with version): SQL Server 2019

ADO.NET Provider (with version):  Microsoft.Data.SqlClient 5.20

Operating system: Windows 10

.NET Version: .NET Core 6

### PR DESCRIPTION
**Pull request by igor-tkachev**
Fixes #4476.

### CODE REVIEW

**Review by viceroypenguin — APPROVED** [review]

File: Source/LinqToDB/DataOptions.cs [hunk 1]
```diff
@@ -94,19 +94,19 @@ public override IEnumerable<IOptionSet> OptionSets
...
 
 			if (type == typeof(LinqOptions))        return (TSet?)(IOptionSet?)LinqOptions;
 			if (type == typeof(RetryPolicyOptions)) return (TSet?)(IOptionSet?)RetryPolicyOptions;
 			if (type == typeof(ConnectionOptions))  return (TSet?)(IOptionSet?)ConnectionOptions;
-			if (type == typeof(DataContextOptions)) return (TSet?)(IOptionSet?)_dataContextOptions;
-			if (type == typeof(BulkCopyOptions))    return (TSet?)(IOptionSet?)_bulkCopyOptions;
-			if (type == typeof(SqlOptions))         return (TSet?)(IOptionSet?)_sqlOptions;
+			if (type == typeof(DataContextOptions)) return (TSet?)(IOptionSet?)(createKnownOptions ? DataContextOptions : _dataContextOptions);
+			if (type == typeof(BulkCopyOptions))    return (TSet?)(IOptionSet?)(createKnownOptions ? BulkCopyOptions    : _bulkCopyOptions);
+			if (type == typeof(SqlOptions))         return (TSet?)(IOptionSet?)(createKnownOptions ? SqlOptions         : _sqlOptions);
```
Comment (MaceWindu) [inline Source/LinqToDB/DataOptions.cs:107]: Could you explain difference in handling of first and last 3 options?

Comment (igor-tkachev) [reply Source/LinqToDB/DataOptions.cs:107]: I do not remember.

Comment (igor-tkachev) [reply Source/LinqToDB/DataOptions.cs:107]: I think last three option sets mostly unused, so in most cases we do not need to create it.

**Review by viceroypenguin — APPROVED** [review]

**Review by MaceWindu — APPROVED** [review]

**Review by viceroypenguin — APPROVED** [review]

**Pull request merged**